File: services/lucy-dotnet/src/Lucy.AuthPayment.Api/Modules/Payment/Controllers/PaymentController.cs
using lucy_auth_payment.Modules.Payment.DTOs;
using lucy_auth_payment.Modules.Payment.Services;
using Microsoft.AspNetCore.Mvc;

namespace lucy_auth_payment.Modules.Payment.Controllers;

[ApiController]
[Route("api/payment")]
public class PaymentController : ControllerBase
{
    private readonly IPaymentService _paymentService;

    public PaymentController(IPaymentService paymentService)
    {
        _paymentService = paymentService;
    }

    [HttpGet("wallet")]
    public async Task<IActionResult> GetWallet()
    {
        return Ok(await _paymentService.GetOrCreateWalletAsync(GetCurrentUserId()));
    }

    [HttpGet("transactions")]
    public async Task<IActionResult> GetTransactions()
    {
        return Ok(await _paymentService.GetTransactionsAsync(GetCurrentUserId()));
    }

    [HttpPost("deposit")]
    public async Task<IActionResult> Deposit([FromBody] DepositRequest request)
    {
        return Ok(await _paymentService.DepositAsync(GetCurrentUserId(), request));
    }

    [HttpPost("purchase/content")]
    public async Task<IActionResult> PurchaseContent([FromBody] PurchaseContentRequest request)
    {
        return Ok(await _paymentService.PurchaseContentAsync(GetCurrentUserId(), request));
    }

    [HttpPost("purchase/live")]
    public async Task<IActionResult> PurchaseLive([FromBody] PurchaseLiveRequest request)
    {
        return Ok(await _paymentService.PurchaseLiveAsync(GetCurrentUserId(), request));
    }

    [HttpPost("donate")]
    public async Task<IActionResult> Donate([FromBody] DonateRequest request)
    {
        return Ok(await _paymentService.DonateAsync(GetCurrentUserId(), request));
    }

    [HttpPost("transfer")]
    public async Task<IActionResult> Transfer([FromBody] TransferRequest request)
    {
        var type = request.TransferType.Trim().ToLowerInvariant();
        if (type == "buypodcast")
        {
            return Ok(await _paymentService.PurchaseContentAsync(GetCurrentUserId(), new PurchaseContentRequest
            {
                ContentId = request.RefId ?? request.ToUserId ?? string.Empty
            }));
        }
        if (type == "paylive")
        {
            return Ok(await _paymentService.PurchaseLiveAsync(GetCurrentUserId(), new PurchaseLiveRequest
            {
                RoomId = request.RefId ?? request.ToUserId ?? string.Empty
            }));
        }
        if (string.IsNullOrWhiteSpace(request.ToUserId))
        {
            throw new InvalidOperationException("ToUserId is required for donate transfer.");
        }
        return Ok(await _paymentService.DonateAsync(GetCurrentUserId(), new DonateRequest
        {
            ToUserId = request.ToUserId,
            Amount = request.Amount,
            RoomId = request.RefId,
            MessageText = request.MessageText
        }));
    }

    [HttpPost("withdraw")]
    public async Task<IActionResult> Withdraw([FromBody] WithdrawRequest request)
    {
        return Ok(await _paymentService.WithdrawAsync(GetCurrentUserId(), request));
    }

    [HttpGet("admin/settings/momo")]
    public async Task<IActionResult> GetMomoSetting()
    {
        EnsureAdmin();
        return Ok(await _paymentService.GetMomoSettingAsync());
    }

    [HttpPost("admin/settings/momo")]
    public async Task<IActionResult> SaveMomoSetting([FromBody] SavePaymentSettingRequest request)
    {
        EnsureAdmin();
        return Ok(await _paymentService.SaveMomoSettingAsync(request));
    }

    [HttpPost("admin/settings/momo/qr")]
    public async Task<IActionResult> UploadMomoQr([FromForm] IFormFile file)
    {
        EnsureAdmin();
        return Ok(await _paymentService.UploadMomoQrAsync(file));
    }

    [HttpGet("admin/topup-orders")]
    public async Task<IActionResult> GetTopUpOrders([FromQuery] string? status)
    {
        EnsureAdmin();
        return Ok(await _paymentService.GetTopUpOrdersAsync(status));
    }

    [HttpPost("admin/topup-orders/{topUpOrderId}/approve")]
    public async Task<IActionResult> ApproveTopUpOrder(string topUpOrderId)
    {
        EnsureAdmin();
        return Ok(await _paymentService.ApproveTopUpOrderAsync(topUpOrderId));
    }

    [HttpPost("admin/topup-orders/{topUpOrderId}/reject")]
    public async Task<IActionResult> RejectTopUpOrder(string topUpOrderId, [FromBody] RejectTopUpRequest? request)
    {
        EnsureAdmin();
        return Ok(await _paymentService.RejectTopUpOrderAsync(topUpOrderId, request?.Reason));
    }

    [HttpGet("admin/withdraw-requests")]
    public async Task<IActionResult> GetWithdrawRequests([FromQuery] string? status)
    {
        EnsureAdmin();
        return Ok(await _paymentService.GetWithdrawRequestsAsync(status));
    }

    [HttpPost("admin/withdraw/approve/{withdrawRequestId}")]
    public async Task<IActionResult> ApproveWithdraw(string withdrawRequestId)
    {
        EnsureAdmin();
        return Ok(await _paymentService.ApproveWithdrawAsync(withdrawRequestId));
    }

    [HttpPost("admin/withdraw/reject/{withdrawRequestId}")]
    public async Task<IActionResult> RejectWithdraw(string withdrawRequestId, [FromBody] RejectWithdrawRequest? request)
    {
        EnsureAdmin();
        return Ok(await _paymentService.RejectWithdrawAsync(withdrawRequestId, request?.RejectReason));
    }

    private string GetCurrentUserId()
    {
        if (Request.Headers.TryGetValue("X-User-Id", out var userId) && !string.IsNullOrWhiteSpace(userId))
        {
            return userId.ToString();
        }
        throw new InvalidOperationException("Missing X-User-Id header.");
    }

    private void EnsureAdmin()
    {
        if (!string.Equals(GetCurrentUserId(), "Uadmin", StringComparison.OrdinalIgnoreCase))
        {
            throw new InvalidOperationException("Admin permission is required.");
        }
    }
}

File: lucy-web-stack/lucy-backend/src/main/java/com/lucy/backend/payment/controller/PaymentController.java
package com.lucy.backend.payment.controller;

import com.lucy.backend.payment.dto.*;
import com.lucy.backend.payment.service.IPaymentService;
import jakarta.servlet.http.HttpServletRequest;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;

@RestController
@RequestMapping("/api/payment")
@CrossOrigin(originPatterns = "*")
public class PaymentController {

    private final IPaymentService paymentService;

    public PaymentController(IPaymentService paymentService) {
        this.paymentService = paymentService;
    }

    private String getCurrentUserId(HttpServletRequest request) {
        String userId = null;
        if (org.springframework.security.core.context.SecurityContextHolder.getContext().getAuthentication() != null) {
            userId = org.springframework.security.core.context.SecurityContextHolder.getContext().getAuthentication().getName();
        }
        if (userId == null || userId.trim().isEmpty() || "anonymousUser".equals(userId)) {
            userId = request.getHeader("X-User-Id");
        }
        if (userId == null || userId.trim().isEmpty()) {
            throw new RuntimeException("Missing X-User-Id header.");
        }
        return userId;
    }

    private void ensureAdmin(HttpServletRequest request) {
        if (!"Uadmin".equalsIgnoreCase(getCurrentUserId(request))) {
            throw new RuntimeException("Admin permission is required.");
        }
    }

    @GetMapping("/wallet")
    public ResponseEntity<?> getWallet(HttpServletRequest request) {
        return ResponseEntity.ok(paymentService.getOrCreateWalletAsync(getCurrentUserId(request)));
    }

    @GetMapping("/transactions")
    public ResponseEntity<?> getTransactions(HttpServletRequest request) {
        return ResponseEntity.ok(paymentService.getTransactionsAsync(getCurrentUserId(request)));
    }

    @GetMapping("/gifts")
    public ResponseEntity<?> getGifts() {
        return ResponseEntity.ok(paymentService.getGiftsAsync());
    }

    @PostMapping("/deposit")
    public ResponseEntity<?> deposit(HttpServletRequest request, @RequestBody DepositRequest depositRequest) {
        return ResponseEntity.ok(paymentService.depositAsync(getCurrentUserId(request), depositRequest));
    }

    @PostMapping("/purchase/content")
    public ResponseEntity<?> purchaseContent(HttpServletRequest request, @RequestBody PurchaseContentRequest purchaseRequest) {
        return ResponseEntity.ok(paymentService.purchaseContentAsync(getCurrentUserId(request), purchaseRequest));
    }

    @PostMapping("/purchase/live")
    public ResponseEntity<?> purchaseLive(HttpServletRequest request, @RequestBody PurchaseLiveRequest purchaseRequest) {
        return ResponseEntity.ok(paymentService.purchaseLiveAsync(getCurrentUserId(request), purchaseRequest));
    }

    @PostMapping("/donate")
    public ResponseEntity<?> donate(HttpServletRequest request, @RequestBody DonateRequest donateRequest) {
        return ResponseEntity.ok(paymentService.donateAsync(getCurrentUserId(request), donateRequest));
    }

    @PostMapping("/transfer")
    public ResponseEntity<?> transfer(HttpServletRequest request, @RequestBody TransferRequest transferRequest) {
        String type = transferRequest.getTransferType() != null ? transferRequest.getTransferType().trim().toLowerCase() : "";
        if ("buypodcast".equals(type)) {
            PurchaseContentRequest pcr = new PurchaseContentRequest();
            pcr.setContentId(transferRequest.getRefId() != null ? transferRequest.getRefId() : transferRequest.getToUserId());
            return ResponseEntity.ok(paymentService.purchaseContentAsync(getCurrentUserId(request), pcr));
        }
        if ("paylive".equals(type)) {
            PurchaseLiveRequest plr = new PurchaseLiveRequest();
            plr.setRoomId(transferRequest.getRefId() != null ? transferRequest.getRefId() : transferRequest.getToUserId());
            return ResponseEntity.ok(paymentService.purchaseLiveAsync(getCurrentUserId(request), plr));
        }
        if (transferRequest.getToUserId() == null || transferRequest.getToUserId().trim().isEmpty()) {
            throw new RuntimeException("ToUserId is required for donate transfer.");
        }
        DonateRequest dr = new DonateRequest();
        dr.setToUserId(transferRequest.getToUserId());
        dr.setAmount(transferRequest.getAmount());
        dr.setRoomId(transferRequest.getRefId());
        dr.setMessageText(transferRequest.getMessageText());
        return ResponseEntity.ok(paymentService.donateAsync(getCurrentUserId(request), dr));
    }

    @PostMapping("/withdraw")
    public ResponseEntity<?> withdraw(HttpServletRequest request, @RequestBody WithdrawRequest withdrawRequest) {
        return ResponseEntity.ok(paymentService.withdrawAsync(getCurrentUserId(request), withdrawRequest));
    }

    @GetMapping("/admin/settings/momo")
    public ResponseEntity<?> getMomoSetting(HttpServletRequest request) {
        ensureAdmin(request);
        return ResponseEntity.ok(paymentService.getMomoSettingAsync());
    }

    @GetMapping("/settings/momo")
    public ResponseEntity<?> getMomoSettingPublic(HttpServletRequest request) {
        return ResponseEntity.ok(paymentService.getMomoSettingAsync());
    }

    @PostMapping("/admin/settings/momo")
    public ResponseEntity<?> saveMomoSetting(HttpServletRequest request, @RequestBody SavePaymentSettingRequest saveRequest) {
        ensureAdmin(request);
        return ResponseEntity.ok(paymentService.saveMomoSettingAsync(saveRequest));
    }

    @PostMapping("/admin/settings/momo/qr")
    public ResponseEntity<?> uploadMomoQr(HttpServletRequest request, @RequestParam("file") MultipartFile file) {
        ensureAdmin(request);
        return ResponseEntity.ok(paymentService.uploadMomoQrAsync(file));
    }

    @GetMapping("/admin/topup-orders")
    public ResponseEntity<?> getTopUpOrders(HttpServletRequest request, @RequestParam(value = "status", required = false) String status) {
        ensureAdmin(request);
        return ResponseEntity.ok(paymentService.getTopUpOrdersAsync(status));
    }

    @PostMapping("/admin/topup-orders/{topUpOrderId}/approve")
    public ResponseEntity<?> approveTopUpOrder(HttpServletRequest request, @PathVariable String topUpOrderId) {
        ensureAdmin(request);
        return ResponseEntity.ok(paymentService.approveTopUpOrderAsync(topUpOrderId));
    }

    @PostMapping("/admin/topup-orders/{topUpOrderId}/reject")
    public ResponseEntity<?> rejectTopUpOrder(HttpServletRequest request, @PathVariable String topUpOrderId, @RequestBody(required = false) RejectTopUpRequest rejectRequest) {
        ensureAdmin(request);
        String reason = rejectRequest != null ? rejectRequest.getReason() : null;
        return ResponseEntity.ok(paymentService.rejectTopUpOrderAsync(topUpOrderId, reason));
    }

    @GetMapping("/admin/withdraw-requests")
    public ResponseEntity<?> getWithdrawRequests(HttpServletRequest request, @RequestParam(value = "status", required = false) String status) {
        ensureAdmin(request);
        return ResponseEntity.ok(paymentService.getWithdrawRequestsAsync(status));
    }

    @PostMapping("/admin/withdraw/approve/{withdrawRequestId}")
    public ResponseEntity<?> approveWithdraw(HttpServletRequest request, @PathVariable String withdrawRequestId) {
        ensureAdmin(request);
        return ResponseEntity.ok(paymentService.approveWithdrawAsync(withdrawRequestId));
    }

    @PostMapping("/admin/withdraw/reject/{withdrawRequestId}")
    public ResponseEntity<?> rejectWithdraw(HttpServletRequest request, @PathVariable String withdrawRequestId, @RequestBody(required = false) RejectWithdrawRequest rejectRequest) {
        ensureAdmin(request);
        String reason = rejectRequest != null ? rejectRequest.getRejectReason() : null;
        return ResponseEntity.ok(paymentService.rejectWithdrawAsync(withdrawRequestId, reason));
    }
}
